### PR TITLE
Beacon: Improve validation of new group's public key

### PR DESCRIPTION
**Pull request by pdyraga**
~~Depends on #3029~~

Instead of checking the uniqueness in `approveDkgResult`, when the non-unique
group passed the challenge period, we should check it in `submitDkgResult`. This
way, in an extremely unlikely case, the same group is getting chosen and they
produce the same public key, DKG will not be blocked forever.

The same fix has been applied to ECDSA `WalletRegistry`, see #3029.

This change increases the gas cost of `submitDkgResult` by ~2640 gas units on
average.

Before/after:
```
| submitDkgResult   ·  232349  ·  239837  ·  235096  │
| submitDkgResult   ·  234996  ·  242484  ·  237743  │
| approveDkgResult  ·  339712  ·  418859  ·  363560  │
| approveDkgResult  ·  339473  ·  418620  ·  363321  |
```

### CODE REVIEW

File: solidity/random-beacon/contracts/libraries/Groups.sol [hunk 1]
```diff
@@ -65,16 +85,6 @@ library Groups {
...
 
-        require(
-            self.groupsData[groupPubKeyHash].registrationBlockNumber == 0,
-            "Group with this public key was already registered"
-        );
-
-        require(
-            self.groupsRegistry.length <= type(uint64).max,
-            "Max number of registered groups reached"
-        );
```
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/Groups.sol:76]: I sacrificed this check to not run into contract size limit problems. Even if we were adding one new group per second, [2^64 is more than the age of Earth](https://www.wolframalpha.com/input?i=2**64+seconds+to+years).

**Comment by dimpar** [comment]
no more comments

File: solidity/random-beacon/contracts/libraries/Groups.sol [hunk 2]
```diff
@@ -51,6 +51,26 @@ library Groups {
...
+        view
+    {
+        bytes32 groupPubKeyHash = keccak256(groupPubKey);
+
+        require(
+            self.groupsData[groupPubKeyHash].registrationBlockNumber == 0,
+            "Group with this public key was already registered"
+        );
+
+        require(groupPubKey.length == 128, "Invalid length of the public key");
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/Groups.sol]: Good point, changed also for `WalletRegistry`: https://github.com/keep-network/keep-core/pull/3030/commits/6a0e800d7789104d842bd6848f4be2e68d0b1bd7

File: solidity/random-beacon/contracts/RandomBeacon.sol [hunk 3]
```diff
@@ -465,7 +465,7 @@ contract RandomBeacon is IRandomBeacon, IApplication, Governable, Reimbursable {
         // Gas parameters were adjusted based on Ethereum state in April 2022.
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:465]: They were adjusted in April 2022 though, I was not touching them when working on this PR.

I think having a specific date makes sense - it provides a information about exact moment in time and a reference to circumstances based on which initial parameters were set.

[the remark this answers is not in the document]
Comment (dimpar) [reply solidity/random-beacon/contracts/RandomBeacon.sol:465]: `_dkgResultSubmissionGas` was changed in this PR, but the change is not that big and the offsets are the same, so I guess we can leave it as is just like you said.

**Review by dimpar — APPROVED** [review]

**Pull request merged**